### ISSUE TITLE
Error with LSTMTorch

### ISSUE THREAD

**Issue by gcelano** [ufal/evalatin2024-latinpipe#2]
I am getting the following error when trying to train:

```
Traceback (most recent call last):
  File "/home/celano/Documents/evalatin2024-latinpipe/latinpipe_evalatin24.py", line 936, in <module>
    main([] if "__file__" not in globals() else None)
  File "/home/celano/Documents/evalatin2024-latinpipe/latinpipe_evalatin24.py", line 867, in main
    model = LatinPipeModel(train, args)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/celano/Documents/evalatin2024-latinpipe/latinpipe_evalatin24.py", line 669, in __init__
    hidden = getattr(self, args.rnn_type)(args.rnn_dim)(embeddings, keras.ops.sum(inputs[1][..., 0] > -1, axis=-1))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/celano/.conda/envs/evalatin/lib/python3.11/site-packages/keras/src/utils/traceback_utils.py", line 122, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/home/celano/.conda/envs/evalatin/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/celano/.conda/envs/evalatin/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/celano/Documents/evalatin2024-latinpipe/latinpipe_evalatin24.py", line 543, in call
    packed_result, _ = self._lstm(torch.nn.utils.rnn.pack_padded_sequence(inputs, lengths.cpu(), batch_first=True, enforce_sorted=False))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/celano/.conda/envs/evalatin/lib/python3.11/site-packages/optree/ops.py", line 747, in tree_map
    return treespec.unflatten(map(func, *flat_args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/celano/.conda/envs/evalatin/lib/python3.11/site-packages/torch/nn/utils/rnn.py", line 67, in __new__
    *_packed_sequence_init_args(data, batch_sizes, sorted_indices,
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/celano/.conda/envs/evalatin/lib/python3.11/site-packages/torch/nn/utils/rnn.py", line 183, in _packed_sequence_init_args
    assert isinstance(data, (list, tuple)) and len(data) == 2
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

AssertionError: Exception encountered when calling LSTMTorch.call().

Could not automatically infer the output shape / dtype of 'lstm_torch' (of type LSTMTorch). Either the `LSTMTorch.call()` method is incorrect, or you need to implement the `LSTMTorch.compute_output_spec() / compute_output_shape()` method. Error encountered:

Arguments received by LSTMTorch.call():
  • args=('<KerasTensor shape=(None, None, 768), dtype=float32, sparse=False, name=keras_tensor_4>', '<KerasTensor shape=(None,), dtype=int32, sparse=False, name=keras_tensor_19>')
  • kwargs=<class 'inspect._empty'>
```

**Comment by foxik** [comment]
Sorry for long reaction, I am currently on vacation.

The problem was caused by an update in Keras 3.4, I have updated the code so that it works with Keras 3.0-3.5 (the latest one).

BTW, we will release the model from EvaLatin24 when I get back (sometimes in September).

**Comment by locusclassicus** [comment]
It seems there is also a compatibility issue with newer versions of torch and cuda. If I use `torch 2.0.1+cu118`, there is an error `module 'torch' has no attribute 'float8_e4m3fn` (discussed [here](https://stackoverflow.com/questions/78929592/torch-has-no-attribute-float8-e4m3fn)). An upgrade to newer versions of `torch`, however, leads to the error of compatibility with `ufal.chu-liu-edmonds`. 

I have been upgrading and downgrading various packages to make it work, but for no good at all.

**Comment by foxik** [comment]
Hi, @locusclassicus,

the problem you are facing is not dependent on CUDA versions, it is caused by the fact that `keras` needs newer PyTorch (the version 2.0.1 you have is quite old); I would use the newest Keras plus the newest PyTorch (so 3.5 Keras and 2.4 PyTorch).
- For reference, https://github.com/keras-team/keras/blob/master/requirements-torch-cuda.txt is the list of packages Keras suggests for torch backend, and it exactly specifies 2.4.

The CUDA versions are an independent issue. Depending on your system you can choose a suitable CUDA version -- here https://pytorch.org/get-started/locally/ you can find commands for installing 2.4 PyTorch with CUDA 11.8, 12.1, 12.4. Note that the minor CUDA versions are backward compatible, so you want to install PyTorch built with largest CUDA which is smaller or equal to your installed version (i.e., if you have 12.2 installed, you should install PyTorch with CUDA 12.1; BTW, that is currently the default, if you just say `pip install torch`).

In other words, I believe both of these issues should be solved by creating a fresh virtual environment and then running `VENV_DIR/bin/pip install -r requirements.txt` -- if you do that, do the problems still persist?

> An upgrade to newer versions of `torch`, however, leads to the error of compatibility with `ufal.chu-liu-edmonds`.

If that is the case, then that is definitely a bug. However, I am not sure how it could happen, because `ufal.chu-liu-edmonds` does not use `torch` at all, it only uses `numpy`, and even the newest `numpy==2.1.0` seems to work fine. What is the error you are seeing with the newer version of `torch`?

**Comment by locusclassicus** [comment]
Hi @foxik , thank you so much for the clarification! I should have thought myself of a fresh environment, but apparently my Latin is better than my Python 😳 Anyway the model is already learning, and I look forward to a new Latin parser 😊 Many thanks again for your help and for the script of course!
